src/nspawn/nspawn.c

```diff
@@ -2226,97 +2227,120 @@ static bool should_enable_fuse(void) {
-                                r = touch(to);
-                                if (r < 0)
-                                        return log_error_errno(r, "touch (%s) failed: %m", to);
-                                r = mount_nofollow_verbose(LOG_DEBUG, from, to, NULL, MS_BIND, NULL);
-                                if (r < 0)
-                                        return log_error_errno(r, "Both mknod and bind mount (%s) failed: %m", to);
-                        } else {
-                                r = userns_lchown(to, 0, 0);
-                                if (r < 0)
-                                        return log_error_errno(r, "chown() of device node %s failed: %m", to);
-                        }
+        _cleanup_free_ char *t = path_join("..", node);
+        if (!t)
+                return log_oom();
 
-                        dn = path_join("/dev", S_ISCHR(st.st_mode) ? "char" : "block");
-                        if (!dn)
-                                return log_oom();
+        if (symlink(t, prefixed) < 0)
+                log_debug_errno(errno, "Failed to symlink '%s' to '%s', ignoring: %m", t, prefixed);
 
-                        r = userns_mkdir(dest, dn, 0755, 0, 0);
-                        if (r < 0)
-                                return log_error_errno(r, "Failed to create '%s': %m", dn);
+        return 0;
+}
 
-                        if (asprintf(&sl, "%s/%u:%u", dn, major(st.st_rdev), minor(st.st_rdev)) < 0)
-                                return log_oom();
+static int copy_devnodes(const char *dest, bool enable_fuse) {
+        int r = 0;
 
-                        prefixed = path_join(dest, sl);
-                        if (!prefixed)
-                                return log_oom();
+        assert(dest);
 
-                        t = path_join("..", *d);
-                        if (!t)
-                                return log_oom();
+        FOREACH_STRING(node, "null", "zero", "full", "random", "urandom", "tty") {
+                r = copy_devnode_one(dest, node, /* ignore_mknod_failure = */ false);
+                if (r < 0)
+                        return r;
+        }
 
-                        if (symlink(t, prefixed) < 0)
-                                log_debug_errno(errno, "Failed to symlink '%s' to '%s': %m", t, prefixed);
-                }
+        if (enable_fuse) {
+                r = copy_devnode_one(dest, "fuse", /* ignore_mknod_failure = */ false);
+                if (r < 0)
+                        return r;
         }
 
-        return r;
+        /* We unconditionally try to create /dev/net/tun, but let's ignore failure if --private-network is
+         * unspecified. The failure can be triggered when e.g. DevicePolicy= is set, but DeviceAllow= does
+         * not contains the device node, and --private-users=pick is specified. */
+        r = copy_devnode_one(dest, "net/tun", /* ignore_mknod_failure = */ !arg_private_network);
```

Comment:
Should the condition not rather be `arg_uid_shift != 0 && !arg_private_network`? Clearly, when `--private-network` is given, the container may want to open it as indicated in the comment. But also when there is no uid shifting going on, the container may inherit capabilities for configuring network interfaces in the initial network namespace and want to create tun/tap interfaces there. Also when not shifting uids, we expect `mknod` to not fail with `-EPERM`.